StyleCop.Analyzers/StyleCop.Analyzers/Helpers/SymbolNameHelpers.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) Tunnel Vision Laboratories, LLC. All Rights Reserved.
+// Licensed under the Apache License, Version 2.0. See LICENSE in the project root for license information.
+
+namespace StyleCop.Analyzers.Helpers
+{
+    using System;
+    using System.Text;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using StyleCop.Analyzers.Lightup;
+
+    /// <summary>
+    /// Contains helper methods to work with symbol names consistently over different C# versions.
+    /// </summary>
+    internal static class SymbolNameHelpers
+    {
+        private const string GenericTypeParametersOpen = "<";
+        private const string GenericTypeParametersClose = ">";
+        private const string GenericSeparator = ", ";
+
+        /// <summary>
+        /// Generates the qualified name for the given symbol.
+        /// </summary>
+        /// <param name="symbol">The symbol to use.</param>
+        /// <returns>The generated qualified name.</returns>
+        public static string ToQualifiedString(this ISymbol symbol)
+        {
+            var builder = ObjectPools.StringBuilderPool.Allocate();
+
+            if (symbol is INamedTypeSymbol namedTypeSymbol)
+            {
+                if (SpecialTypeHelper.TryGetPredefinedType(namedTypeSymbol.SpecialType, out PredefinedTypeSyntax specialTypeSyntax))
+                {
+                    return specialTypeSyntax.ToFullString();
+                }
+
+                if (namedTypeSymbol.IsTupleType())
+                {
+                    namedTypeSymbol = namedTypeSymbol.TupleUnderlyingType();
+                }
+
+                AppendQualifiedSymbolName(builder, namedTypeSymbol);
+
+                if (namedTypeSymbol.IsGenericType)
+                {
+                    builder.Append(GenericTypeParametersOpen);
+
+                    foreach (var typeArgument in namedTypeSymbol.TypeArguments)
+                    {
+                        if (typeArgument is INamedTypeSymbol namedTypeArgument && typeArgument.IsTupleType())
```

Comment:
💡 I believe full coverage of this condition would need a test involving an array type as a generic type argument.

Reply:
I'll add that 👍